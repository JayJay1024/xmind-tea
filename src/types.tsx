export enum EnumBill { Out=0, In=1 };  // 支出收入
export enum EnumMonth { Jan, Fed, Mar, Apr, May, Jun, Jul, Aug, Seq, Oct, Nov, Dec };

// 账单
export type TypeBill = {
    time: Date,
    type: EnumBill,
    category?: string,
    amount: number
};

// 分类
export type TypeCategory = {
    id: string,
    name: string,
    type: EnumBill
}

// 收支
export type TypeAmountInOut = {
    in: number,
    out: number
}

export const ConstMonthsText = ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月'];
