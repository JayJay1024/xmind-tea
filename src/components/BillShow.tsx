import React, { useState, useEffect } from 'react';
import {
    Table,
    Menu, 
    Dropdown,
    Typography,
} from 'antd';
import { DownOutlined } from '@ant-design/icons';
import dayjs from 'dayjs';

import {
    EnumBill,
    EnumMonth,
    TypeBill,
    TypeCategory,
    TypeAmountInOut,
    ConstMonthsText
} from '../types';
import '../styles.css';

type TypeProp = {
    dataBill: TypeBill[],
    dataCategory: TypeCategory[],
    monthIndex: EnumMonth,
    onClick: (o: object) => void
}

const { Text } = Typography;

const BillShow: React.FunctionComponent<TypeProp> = (props): JSX.Element => {
    const [amountInOut, setAmountInOut]   = useState<TypeAmountInOut>({in: 0, out: 0});  // 选择月份后收支统计

    let billColumns = [{
        title: '账单时间',
        dataIndex: 'time',
        key: 'time',
        align: 'center' as any,
    }, {
        title: '账单类型',
        dataIndex: 'type',
        key: 'type',
        align: 'center' as any,
    }, {
        title: '账单分类',
        dataIndex: 'category',
        key: 'category',
        align: 'center' as any,
        filters: props.dataCategory.map(v => { return { text: v.name, value: v.name } }),
        onFilter: (value: any, record: any) => record.category.indexOf(value as string) === 0,
    }, {
        title: '账单金额(￥)',
        dataIndex: 'amount',
        key: 'amount',
        align: 'center' as any,
    }];

    const checkBillDataForMonth = (): object[] => {
        return props.dataBill.filter(v => Number(dayjs(v.time).month()) === Number(props.monthIndex)).map((v, i) => {
            const c = props.dataCategory.find(vc => vc.id === v.category);
            return {
                key: i,
                time: dayjs(v.time).format(),
                type: v.type === EnumBill.In? "收入" : "支出",
                category: c ? c.name : v.category,
                amount: v.amount.toFixed(2)
            }
        });
    }

    // 月份的下拉菜单选项
    const MonthMenu: React.ReactElement = (
        <Menu onClick={props.onClick} >
            {ConstMonthsText.map((v, i) => (<Menu.Item key={i}>{v}</Menu.Item>))}
        </Menu>
    );

    // 计算收支
    useEffect(() => {
        let amountIO: TypeAmountInOut = {in: 0, out: 0};
        props.dataBill.filter(v => Number(dayjs(v.time).month()) === Number(props.monthIndex)).forEach(v => {
            v.type === EnumBill.In ? amountIO.in += v.amount : amountIO.out += v.amount;
        });
        setAmountInOut(amountIO);
    }, [props.monthIndex, props.dataBill]);

    return (
        <div className="bill-show">
            <div className="monnth-select-in-out">
                <Dropdown overlay={MonthMenu}>
                    <a className="ant-dropdown-link" onClick={e => e.preventDefault()}>
                        {ConstMonthsText[props.monthIndex]} <DownOutlined />
                    </a>
                </Dropdown>
                <Text className="in-out">收入总额: <Text code>￥{amountInOut.in.toFixed(2)} </Text></Text>
                <Text className="in-out">支出总额: <Text code>￥{amountInOut.out.toFixed(2)}</Text></Text>
            </div>
            <Table
                pagination={false}
                scroll={{ y: 420 }}
                loading={props.dataCategory.length ? false : true}
                columns={billColumns}
                dataSource={checkBillDataForMonth()}
            />
        </div>
    );
};

export default BillShow;
