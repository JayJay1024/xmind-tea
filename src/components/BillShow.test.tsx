import React from 'react';
import { RenderResult, render, act, screen } from '@testing-library/react';
import '../matchMedia.tsx';

import { EnumBill, EnumMonth, TypeBill, TypeCategory } from '../types';
import BillShow from './BillShow';

describe('<BillShow />', () => {
    const fakeDataCategory: TypeCategory[] = [{
        id: 'cid_1111',
        name: 'cname_1111',
        type: EnumBill.In,
    }, {
        id: 'cid_2222',
        name: 'cname_2222',
        type: EnumBill.Out,
    }, {
        id: 'cid_3333',
        name: 'cname_3333',
        type: EnumBill.Out,
    }];
    const fakeDataBill: TypeBill[] = [{
        time: new Date('2048-12-12 11:10:20'),
        type: EnumBill.In,  // 收入，将不计算
        category: 'cid_1111',
        amount: -10,
    }, {
        time: new Date('2048-12-12 11:10:20'),
        type: EnumBill.Out,
        category: 'cid_2222',
        amount: 10,
    }, {
        time: new Date('2048-12-12 12:10:20'),
        type: EnumBill.Out,
        category: 'cid_2222',
        amount: 20,
    }, {
        time: new Date('2048-12-12 13:10:20'),
        type: EnumBill.Out,
        category: 'cid_3333',
        amount: 60,
    }, {
        time: new Date('2048-12-12 14:10:20'),
        type: EnumBill.Out,
        category: 'cid_3333',
        amount: 40,
    }];

    test('January no data', () => {
        // 根据 fakeDataBill ，事实只有 12 月有数据显示
        act(() => {
            render(<BillShow dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Jan} onClick={() => {}} />);  // 1月
        });
        expect(screen.queryByText('cname_1111')).toBeNull();
        expect(screen.queryByText('cname_2222')).toBeNull();
        expect(screen.queryByText('cname_3333')).toBeNull();
    });

    test('December should have data', () => {
        // 根据 fakeDataBill ，事实只有 12 月有数据显示
        act(() => {
            render(<BillShow dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Dec} onClick={() => {}} />);  // 12 月
        });
        expect(screen.queryAllByText('cname_1111').length).toBeGreaterThan(0);
        expect(screen.queryAllByText('cname_2222').length).toBeGreaterThan(0);
        expect(screen.queryAllByText('cname_3333').length).toBeGreaterThan(0);
    });

    test('income and outgo calculate correctly', () => {
        act(() => {
            render(<BillShow dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Dec} onClick={() => {}} />);
        });
        expect(screen.queryByText('￥-10.00')).not.toBeNull();  // 支出统计
        expect(screen.queryByText('￥130.00')).not.toBeNull();  // 收入统计
    });

    test('should match snapshot', () => {
        let renderResult: any = null;
        act(() => {
            renderResult = render(<BillShow dataBill={fakeDataBill} dataCategory={fakeDataCategory} monthIndex={EnumMonth.Dec} onClick={() => {}} />);
        });
        expect(renderResult.asFragment()).toMatchSnapshot();
    });
});
