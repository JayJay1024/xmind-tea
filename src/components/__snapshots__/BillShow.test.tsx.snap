// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BillShow /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="bill-show"
  >
    <div
      class="monnth-select-in-out"
    >
      <a
        class="ant-dropdown-link ant-dropdown-trigger"
      >
        12月 
        <span
          aria-label="down"
          class="anticon anticon-down"
          role="img"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </a>
      <span
        class="ant-typography in-out"
      >
        收入总额: 
        <span
          class="ant-typography"
        >
          <code>
            ￥-10.00 
          </code>
        </span>
      </span>
      <span
        class="ant-typography in-out"
      >
        支出总额: 
        <span
          class="ant-typography"
        >
          <code>
            ￥130.00
          </code>
        </span>
      </span>
    </div>
    <div
      class="ant-table-wrapper"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-fixed-header"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-header"
                style="overflow: hidden;"
              >
                <table
                  style="table-layout: fixed; visibility: hidden;"
                >
                  <colgroup />
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        账单时间
                      </th>
                      <th
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        账单类型
                      </th>
                      <th
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        <div
                          class="ant-table-filter-column"
                        >
                          <span
                            class="ant-table-filter-column-title"
                          >
                            账单分类
                          </span>
                          <span
                            class="ant-table-filter-trigger-container"
                          >
                            <span
                              class="ant-table-filter-trigger ant-dropdown-trigger"
                              role="button"
                              tabindex="-1"
                            >
                              <span
                                aria-label="filter"
                                class="anticon anticon-filter"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  class=""
                                  data-icon="filter"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M349 838c0 17.7 14.2 32 31.8 32h262.4c17.6 0 31.8-14.3 31.8-32V642H349v196zm531.1-684H143.9c-24.5 0-39.8 26.7-27.5 48l221.3 376h348.8l221.3-376c12.1-21.3-3.2-48-27.7-48z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </th>
                      <th
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        账单金额(￥)
                      </th>
                    </tr>
                  </thead>
                </table>
              </div>
              <div
                class="ant-table-body"
                style="overflow-y: scroll; max-height: 420px;"
              >
                <table
                  style="table-layout: fixed;"
                >
                  <colgroup />
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      aria-hidden="true"
                      class="ant-table-measure-row"
                      style="height: 0px;"
                    >
                      <td
                        style="padding: 0px; border: 0px; height: 0px;"
                      />
                      <td
                        style="padding: 0px; border: 0px; height: 0px;"
                      />
                      <td
                        style="padding: 0px; border: 0px; height: 0px;"
                      />
                      <td
                        style="padding: 0px; border: 0px; height: 0px;"
                      />
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="0"
                    >
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        2048-12-12T11:10:20+08:00
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        收入
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        cname_1111
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        -10.00
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="1"
                    >
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        2048-12-12T11:10:20+08:00
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        支出
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        cname_2222
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        10.00
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="2"
                    >
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        2048-12-12T12:10:20+08:00
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        支出
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        cname_2222
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        20.00
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="3"
                    >
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        2048-12-12T13:10:20+08:00
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        支出
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        cname_3333
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        60.00
                      </td>
                    </tr>
                    <tr
                      class="ant-table-row ant-table-row-level-0"
                      data-row-key="4"
                    >
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        2048-12-12T14:10:20+08:00
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        支出
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        cname_3333
                      </td>
                      <td
                        class="ant-table-cell"
                        style="text-align: center;"
                      >
                        40.00
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
